File: reactive-tui-ts/src/examples/tui-demos/plugin_demo.ts
#!/usr/bin/env bun
/**
 * Plugin System Demo - TypeScript Version
 * 
 * Demonstrates the extensible plugin architecture with custom widgets,
 * event interception, and dynamic plugin loading/unloading.
 */

import { createApp, div, text } from '../../packages/tui-bun/src';
import { 
    PluginManager, 
    Plugin, 
    WidgetPlugin, 
    PluginContext,
    PluginEvent,
    PluginResponse,
    PluginCapability,
    type PluginMetadata,
    type WidgetConfig 
} from '../../packages/tui-bun/src/plugin';
import { panel } from '../../packages/tui-bun/src/widgets/panel';
import { bar, barItem } from '../../packages/tui-bun/src/widgets/bar';
import { grid } from '../../packages/tui-bun/src/widgets/grid';

/**
 * Custom gauge widget plugin
 */
class GaugeWidgetPlugin extends WidgetPlugin {
    readonly id = 'gauge-widget';
    readonly widgetType = 'gauge';
    
    readonly metadata: PluginMetadata = {
        name: 'Gauge Widget',
        version: '1.0.0',
        author: 'Demo Author',
        description: 'A circular gauge widget for displaying values',
        dependencies: [],
        capabilities: [PluginCapability.WidgetProvider],
        tags: ['widget', 'gauge', 'visualization'],
    };

    get configSchema() {
        return {
            type: 'object',
            properties: {
                value: { type: 'number', minimum: 0, maximum: 100 },
                label: { type: 'string' },
                color: { type: 'string' },
                size: { type: 'string', enum: ['small', 'medium', 'large'] },
            },
            required: ['value'],
        };
    }

    async initialize(context: PluginContext): Promise<void> {
        console.log('🔌 Gauge Widget Plugin initialized');
        context.setState('theme', 'default');
    }

    async cleanup(): Promise<void> {
        console.log('🔌 Gauge Widget Plugin cleaned up');
    }

    handleEvent(event: PluginEvent): PluginResponse | null {
        if (event.type === 'custom' && event.eventType === 'theme-change') {
            console.log('🎨 Gauge widget received theme change event');
            return { type: 'continue' };
        }
        return null;
    }

    createInstance(config: WidgetConfig): WidgetPlugin {
        return new GaugeWidgetPlugin(config);
    }

    render() {
        const { value = 0, label = 'Gauge', size = 'medium' } = this.config.properties;
        
        // Create a simple text-based gauge representation
        const percentage = Math.min(100, Math.max(0, value)) / 100 * 10;
        const filled = '█'.repeat(Math.floor(percentage));
        const empty = '░'.repeat(10 - Math.floor(percentage));
        
        return div()
            .class('gauge-widget')
            .classes(this.config.cssClasses)
            .child(text(`${label}: [${filled}${empty}] ${value.toFixed(0)}%`));
    }
}

/**
 * Event logger plugin that intercepts all events
 */
class EventLoggerPlugin extends Plugin {
    readonly id = 'event-logger';
    private eventCount = 0;
    
    readonly metadata: PluginMetadata = {
        name: 'Event Logger',
        version: '1.0.0',
        author: 'Demo Author',
        description: 'Logs all plugin events for debugging',
        dependencies: [],
        capabilities: [PluginCapability.EventInterceptor],
        tags: ['debug', 'logging'],
    };

    async initialize(context: PluginContext): Promise<void> {
        console.log('📝 Event Logger Plugin initialized');
    }

    async cleanup(): Promise<void> {
        console.log(`📝 Event Logger Plugin cleaned up. Total events logged: ${this.eventCount}`);
    }

    handleEvent(event: PluginEvent): PluginResponse | null {
        this.eventCount++;
        
        switch (event.type) {
            case 'plugin-loaded':
                console.log(`📝 [EVENT] Plugin loaded: ${event.pluginId}`);
                break;
            case 'plugin-unloading':
                console.log(`📝 [EVENT] Plugin unloading: ${event.pluginId}`);
                break;
            case 'widget-created':
                console.log(`📝 [EVENT] Widget created: ${event.widgetType} (${event.widgetId})`);
                break;
            case 'custom':
                console.log(`📝 [EVENT] Custom event: ${event.eventType}`);
                break;
        }
        
        return { type: 'continue' };
    }
}

/**
 * Theme provider plugin
 */
class ThemeProviderPlugin extends Plugin {
    readonly id = 'theme-provider';
    
    readonly metadata: PluginMetadata = {
        name: 'Theme Provider',
        version: '1.0.0',
        author: 'Demo Author',
        description: 'Provides themes for the application',
        dependencies: [],
        capabilities: [PluginCapability.ThemeProvider],
        tags: ['theme', 'styling'],
    };

    private themes = {
        light: {
            background: '#ffffff',
            foreground: '#000000',
            primary: '#0066cc',
        },
        dark: {
            background: '#1a1a1a',
            foreground: '#ffffff',
            primary: '#66b3ff',
        },
    };

    async initialize(context: PluginContext): Promise<void> {
        console.log('🎨 Theme Provider Plugin initialized');
        context.setState('available-themes', Object.keys(this.themes));
        context.setState('current-theme', 'light');
    }

    async cleanup(): Promise<void> {
        console.log('🎨 Theme Provider Plugin cleaned up');
    }

    handleEvent(event: PluginEvent): PluginResponse | null {
        if (event.type === 'custom' && event.eventType === 'get-theme') {
            const themeName = event.data?.theme || 'light';
            return {
                type: 'data',
                data: this.themes[themeName] || this.themes.light,
            };
        }
        return null;
    }
}

async function main() {
    console.log('🔌 Plugin System Demo');
    console.log('====================\n');
    
    // Create plugin manager
    const pluginManager = new PluginManager();
    
    // Register plugins
    await pluginManager.register(new EventLoggerPlugin());
    await pluginManager.register(new ThemeProviderPlugin());
    await pluginManager.register(new GaugeWidgetPlugin());
    
    // State
    let gaugeValues = [75, 42, 89];
    const plugins = pluginManager.listPlugins();
    
    // Create demo app
    const app = createApp({
        title: 'Plugin System Demo',
        component: () => {
            return div()
                .child(
                    grid()
                        .columns(['1fr', '2fr'])
                        .gap(2)
                        .child(
                            panel()
                                .title('Loaded Plugins')
                                .borderStyle('rounded')
                                .child(
                                    div()
                                        .style({ padding: '1rem' })
                                        .children(
                                            plugins.map(plugin => 
                                                div()
                                                    .style({ marginBottom: '1rem' })
                                                    .child(text(`📦 ${plugin.name}`))
                                                    .child(text(`   Version: ${plugin.version}`))
                                                    .child(text(`   Author: ${plugin.author}`))
                                                    .child(text(`   Capabilities: ${plugin.capabilities.join(', ')}`))
                                            )
                                        )
                                )
                        )
                        .child(
                            panel()
                                .title('Custom Widgets Demo')
                                .borderStyle('double')
                                .child(
                                    div()
                                        .style({ padding: '1rem' })
                                        .child(text('Gauge Widgets:'))
                                        .child(text(''))
                                        .children(
                                            gaugeValues.map((value, i) => {
                                                try {
                                                    const widget = pluginManager.createWidget('gauge', {
                                                        id: `gauge-${i}`,
                                                        widgetType: 'gauge',
                                                        properties: {
                                                            value,
                                                            label: `Metric ${i + 1}`,
                                                            size: 'medium',
                                                        },
                                                        cssClasses: ['gauge'],
                                                        eventHandlers: {},
                                                    });
                                                    return widget.render();
                                                } catch (error) {
                                                    return text('Failed to create gauge');
                                                }
                                            })
                                        )
                                        .child(text(''))
                                        .child(text('Plugin Features:'))
                                        .child(text('• Dynamic widget registration'))
                                        .child(text('• Event interception and logging'))
                                        .child(text('• Configuration validation'))
                                        .child(text('• Dependency resolution'))
                                        .child(text(''))
                                        .child(text('Available Widget Types:'))
                                        .children(
                                            pluginManager.getWidgetTypes().map(type =>
                                                text(`• ${type}`)
                                            )
                                        )
                                )
                        )
                )
                .child(
                    bar()
                        .position('bottom')
                        .items([
                            barItem('[R] Reload Plugins'),
                            barItem('[T] Trigger Theme Event'),
                            barItem('[U] Update Gauge Values'),
                            barItem('[L] List Widget Schemas'),
                            barItem('[Q] Quit'),
                        ])
                );
        },
        onKeyPress: async (key) => {
            switch (key) {
                case 'q':
                case 'Q':
                    // Cleanup plugins before exit
                    await pluginManager.unregister('event-logger');
                    await pluginManager.unregister('theme-provider');
                    await pluginManager.unregister('gauge-widget');
                    process.exit(0);
                    break;
                    
                case 'r':
                case 'R':
                    console.log('🔄 Reloading plugins...');
                    // In a real implementation, this would reload plugins
                    break;
                    
                case 't':
                case 'T':
                    // Trigger a custom theme change event
                    pluginManager.broadcastEvent({
                        type: 'custom',
                        eventType: 'theme-change',
                        data: { theme: 'dark' },
                    });
                    break;
                    
                case 'u':
                case 'U':
                    // Update gauge values randomly
                    gaugeValues = gaugeValues.map(() => Math.round(Math.random() * 100));
                    break;
                    
                case 'l':
                case 'L':
                    // List widget schemas
                    console.log('\n📋 Widget Schemas:');
                    for (const widgetType of pluginManager.getWidgetTypes()) {
                        const schema = pluginManager.getWidgetSchema(widgetType);
                        console.log(`\n${widgetType}:`);
                        console.log(JSON.stringify(schema, null, 2));
                    }
                    break;
            }
        },
    });
    
    // Run the app
    await app.run();
}

// Run the demo
main().catch(console.error);

File: examples/plugin_demo.rs
/*!
 * Plugin System Demo
 *
 * Demonstrates the extensible plugin architecture with custom widgets,
 * event interception, and dynamic plugin loading/unloading.
 */

use reactive_tui::components::{div, text, Element};
use reactive_tui::plugin::*;
use reactive_tui::prelude::*;
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// Custom gauge widget plugin
struct GaugeWidgetPlugin {
  config: WidgetConfig,
}

impl Plugin for GaugeWidgetPlugin {
  fn id(&self) -> &str {
    "gauge-widget"
  }

  fn metadata(&self) -> PluginMetadata {
    PluginMetadata {
      id: "gauge-widget".to_string(),
      name: "Gauge Widget".to_string(),
      version: "1.0.0".to_string(),
      author: "Demo Author".to_string(),
      description: "A circular gauge widget for displaying values".to_string(),
      entry_point: "gauge_widget.wasm".to_string(),
      homepage: None,
      dependencies: vec![],
      capabilities: vec![PluginCapability::WidgetProvider],
      tags: vec![
        "widget".to_string(),
        "gauge".to_string(),
        "visualization".to_string(),
      ],
    }
  }

  fn initialize(&mut self, context: &mut PluginContext) -> Result<()> {
    println!("🔌 Gauge Widget Plugin initialized");
    context.set_state("theme", "default");
    Ok(())
  }

  fn cleanup(&mut self) -> Result<()> {
    println!("🔌 Gauge Widget Plugin cleaned up");
    Ok(())
  }

  fn handle_event(&mut self, event: &PluginEvent) -> Option<PluginResponse> {
    match event {
      PluginEvent::Custom {
        event_type,
        data: _,
      } if event_type == "theme-change" => {
        println!("🎨 Gauge widget received theme change event");
        Some(PluginResponse::Continue)
      }
      _ => None,
    }
  }

  fn as_any(&self) -> &dyn std::any::Any {
    self
  }

  fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
    self
  }
}

impl WidgetPlugin for GaugeWidgetPlugin {
  fn widget_type(&self) -> &str {
    "gauge"
  }

  fn create_instance(&self, config: WidgetConfig) -> Box<dyn WidgetPlugin> {
    Box::new(GaugeWidgetPlugin { config })
  }

  fn config_schema(&self) -> serde_json::Value {
    serde_json::json!({
        "type": "object",
        "properties": {
            "value": { "type": "number", "minimum": 0, "maximum": 100 },
            "label": { "type": "string" },
            "color": { "type": "string" },
            "size": { "type": "string", "enum": ["small", "medium", "large"] }
        },
        "required": ["value"]
    })
  }

  fn validate_config(&self, config: &WidgetConfig) -> Result<()> {
    if config.properties.get("value").is_none() {
      return Err(TuiError::plugin("Gauge widget requires 'value' property"));
    }
    Ok(())
  }
}

impl Component for GaugeWidgetPlugin {
  fn render(&self) -> Element {
    let value = self
      .config
      .properties
      .get("value")
      .and_then(|v| v.as_f64())
      .unwrap_or(0.0);
    let label = self
      .config
      .properties
      .get("label")
      .and_then(|v| v.as_str())
      .unwrap_or("Gauge");
    let _size = self
      .config
      .properties
      .get("size")
      .and_then(|v| v.as_str())
      .unwrap_or("medium");

    // Create a simple text-based gauge representation
    let percentage = (value.clamp(0.0, 100.0) / 100.0 * 10.0) as usize;
    let filled = "█".repeat(percentage);
    let empty = "░".repeat(10 - percentage);

    div()
      .class("gauge-widget")
      .classes(&self.config.css_classes)
      .child(text(format!("{label}: [{filled}{empty}] {value:.0}%")).build())
      .build()
  }
}

/// Event logger plugin that intercepts all events
struct EventLoggerPlugin {
  event_count: Arc<RwLock<usize>>,
}

impl Plugin for EventLoggerPlugin {
  fn id(&self) -> &str {
    "event-logger"
  }

  fn metadata(&self) -> PluginMetadata {
    PluginMetadata {
      id: "event-logger".to_string(),
      name: "Event Logger".to_string(),
      version: "1.0.0".to_string(),
      author: "Demo Author".to_string(),
      description: "Logs all plugin events for debugging".to_string(),
      entry_point: "event_logger.wasm".to_string(),
      homepage: None,
      dependencies: vec![],
      capabilities: vec![PluginCapability::EventInterceptor],
      tags: vec!["debug".to_string(), "logging".to_string()],
    }
  }

  fn initialize(&mut self, _context: &mut PluginContext) -> Result<()> {
    println!("📝 Event Logger Plugin initialized");
    Ok(())
  }

  fn cleanup(&mut self) -> Result<()> {
    let count = *self.event_count.read().unwrap();
    println!("📝 Event Logger Plugin cleaned up. Total events logged: {count}");
    Ok(())
  }

  fn handle_event(&mut self, event: &PluginEvent) -> Option<PluginResponse> {
    *self.event_count.write().unwrap() += 1;

    match event {
      PluginEvent::PluginLoaded { plugin_id } => {
        println!("📝 [EVENT] Plugin loaded: {plugin_id}");
      }
      PluginEvent::PluginUnloading { plugin_id } => {
        println!("📝 [EVENT] Plugin unloading: {plugin_id}");
      }
      PluginEvent::WidgetCreated {
        widget_type,
        widget_id,
      } => {
        println!("📝 [EVENT] Widget created: {widget_type} ({widget_id})");
      }
      PluginEvent::Custom { event_type, .. } => {
        println!("📝 [EVENT] Custom event: {event_type}");
      }
      _ => {}
    }

    Some(PluginResponse::Continue)
  }

  fn as_any(&self) -> &dyn std::any::Any {
    self
  }

  fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
    self
  }
}

fn main() -> Result<()> {
  println!("🔌 Plugin System Demo");
  println!("====================\n");

  // Create plugin manager
  let plugin_manager = Arc::new(RwLock::new(PluginManager::new()));

  // Register event logger plugin first (as it's an interceptor)
  {
    let manager = plugin_manager.write().unwrap();
    let logger_plugin = EventLoggerPlugin {
      event_count: Arc::new(RwLock::new(0)),
    };
    manager.register(logger_plugin)?;
  }

  // Register gauge widget plugin
  {
    let manager = plugin_manager.write().unwrap();
    let gauge_plugin = GaugeWidgetPlugin {
      config: WidgetConfig {
        id: "gauge-plugin".to_string(),
        widget_type: "gauge".to_string(),
        properties: serde_json::json!({}),
        css_classes: vec![],
        event_handlers: HashMap::new(),
      },
    };
    manager.register_widget(gauge_plugin)?;
  }

  // State for UI
  let plugin_list = Arc::new(RwLock::new(Vec::new()));
  let _gauge_values = Arc::new(RwLock::new([75.0, 42.0, 89.0]));

  // Update plugin list
  {
    let manager = plugin_manager.read().unwrap();
    *plugin_list.write().unwrap() = manager.list_plugins();
  }

  // Demo complete - print plugin information
  println!("🔌 Plugin Demo Complete!");
  println!("======================\n");

  let plugins = plugin_manager.read().unwrap().list_plugins();
  for plugin in &plugins {
    println!("📦 {} v{}", plugin.name, plugin.version);
    println!("   Author: {}", plugin.author);
    println!("   Description: {}", plugin.description);
    println!("   Capabilities: {:?}\n", plugin.capabilities);
  }

  // Create and render some gauge widgets
  println!("🔌 Creating sample widgets:");
  let gauge_values = [75.0, 42.0, 89.0];

  for (i, value) in gauge_values.iter().enumerate() {
    let config = WidgetConfig {
      id: format!("gauge-{i}"),
      widget_type: "gauge".to_string(),
      properties: serde_json::json!({
          "value": value,
          "label": format!("Metric {}", i + 1),
          "size": "medium"
      }),
      css_classes: vec!["gauge".to_string()],
      event_handlers: HashMap::new(),
    };

    match plugin_manager
      .read()
      .unwrap()
      .create_widget("gauge", config)
    {
      Ok(widget) => {
        let element = widget.render();
        println!("✅ Created gauge-{i}: {element:?}");
      }
      Err(e) => println!("❌ Failed to create gauge-{i}: {e}"),
    }
  }

  // Test event broadcasting
  println!("\n🎨 Testing event system:");
  plugin_manager
    .read()
    .unwrap()
    .broadcast_event(PluginEvent::Custom {
      event_type: "theme-change".to_string(),
      data: serde_json::json!({"theme": "dark"}),
    });

  // Cleanup plugins
  {
    let manager = plugin_manager.write().unwrap();
    manager.unregister("event-logger")?;
  }

  Ok(())
}
